pkg/config/descriptor/descriptor.go

```diff
@@ -49,6 +75,167 @@ type finder struct {
+		DisplayName: "RateLimit",
+		Description: "RateLimit By Service",
+		Labels: map[string]dpb.ValueType{
+			"target": dpb.STRING,
+		},
+		RateLimit: true,
+	},
+	manifests: &dpb.AttributeDescriptor{
+		Name:        "target.service",
+		ValueType:   dpb.STRING,
+		Description: "Intended destination of a request",
+	},
+}
+
+// Parse parses a descriptor config into its parts.
+// This entire function is equivalent to jsonpb.Unmarshal()
+// It provides better error reporting about which specific
+// object has the error
+func Parse(cfg string) (dcfg *pb.GlobalConfig, ce *adapter.ConfigErrors) {
+	m := map[string]interface{}{}
+	var err error
+	if err = yaml.Unmarshal([]byte(cfg), &m); err != nil {
+		return nil, ce.Append("DescriptorConfig", err)
+	}
+	var cerr *adapter.ConfigErrors
+	var oarr reflect.Value
+
+	basemsg := map[string]interface{}{}
+	// copy unhandled keys into basemsg
+	for kk, v := range m {
+		if typeMap[dname(kk)] == nil {
+			basemsg[kk] = v
+		}
+	}
+
+	dcfg = &pb.GlobalConfig{}
+
+	ce = ce.Extend(updateMsg("DescriptorConfig", basemsg, dcfg, nil, true))
+
+	//flatten manifest
+	var k dname = manifests
+	val := m[string(k)]
+	if val != nil {
+		mani := []*pb.AttributeManifest{}
+		for midx, msft := range val.([]interface{}) {
+			mname := fmt.Sprintf("%s[%d]", k, midx)
+			manifest := msft.(map[string]interface{})
+			attr := manifest[attributes].([]interface{})
+			delete(manifest, attributes)
+			ma := &pb.AttributeManifest{}
+			if cerr = updateMsg(mname, manifest, ma, typeMap[k], false); cerr != nil {
+				ce = ce.Extend(cerr)
+			}
+			if oarr, cerr = processArray(mname+"."+attributes, attr, typeMap[k]); cerr != nil {
+				ce = ce.Extend(cerr)
+				continue
+			}
+			ma.Attributes = oarr.Interface().([]*dpb.AttributeDescriptor)
+			mani = append(mani, ma)
+		}
+		dcfg.Manifests = mani
+	}
+
+	for k, kfn := range typeMap {
+		if k == manifests {
+			continue
+		}
+		val = m[string(k)]
+		if val == nil {
+			if glog.V(2) {
+				glog.Warningf("%s missing", k)
+			}
+			continue
+		}
+
+		if oarr, cerr = processArray(string(k), val.([]interface{}), kfn); cerr != nil {
+			ce = ce.Extend(cerr)
+			continue
+		}
+		fld := reflect.ValueOf(dcfg).Elem().FieldByName(strings.Title(string(k)))
+		if !fld.IsValid() {
+			continue
+		}
+		fld.Set(oarr)
+	}
+
+	return dcfg, ce
+}
+
+// updateMsg updates a proto.Message using a json message
+// of type []interface{} or map[string]interface{}
+// obj must be previously obtained from a json.Unmarshal
+func updateMsg(ctx string, obj interface{}, dm proto.Message, example proto.Message, allowUnknownFields bool) (ce *adapter.ConfigErrors) {
+	var enc []byte
+	var err error
+
+	if enc, err = json.Marshal(obj); err != nil {
+		return ce.Append(ctx, err)
+	}
+	um := &jsonpb.Unmarshaler{AllowUnknownFields: allowUnknownFields}
+	if err = um.Unmarshal(bytes.NewReader(enc), dm); err != nil {
+		msg := fmt.Sprintf("%v: [%s]", err, string(enc))
+		if example != nil {
+			um := &jsonpb.Marshaler{}
```

Comment:
did you intentionally reuse `um` here?  `um` is earlier defined to be a `jsonpb.Unmarshaler. Might be clearer to call it jsonm or something.